src/reconcile/rec_visibility.c

```diff
@@ -245,11 +249,19 @@ __rec_find_and_save_delete_hs_upd(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_
                 break;
             }
         }
+
+        /* Track the first globally visible self-contained value. */
+        if (!F_ISSET(r, WT_REC_EVICT) && visible_all_upd == NULL &&
+          __wt_txn_upd_visible_all(session, delete_upd) && WT_UPDATE_DATA_VALUE(delete_upd))
+            visible_all_upd = delete_upd;
     }
 
     WT_ASSERT_ALWAYS(session, delete_tombstone == NULL || delete_upd != NULL,
       "If we delete a tombstone from the history store, we must also delete the update.");
 
+    /* Free obsolete updates if exist. */
+    if (!delete_hs_upd_found && visible_all_upd != NULL && visible_all_upd->next != NULL)
+        __wt_update_obsolete_check(session, r->ref, visible_all_upd, false);
```

Comment:
When we need lock and when we don't need it?

Reply:
Other than the reconciliation flow, we need the page lock to check for obsolete updates.